### PR TITLE
Add Parent RecordItem link for easy parent child transaction consolidation

### LINKED ISSUE #3670: Add parent transaction reference logic to parser
### Problem

With the addition of child transactions some transactions no longer exist independently but may also share information relating to a larger process (collection of transactions) e.g. `ContractCall` / `EthereumTransaction` which results in a subsequent `ContractCall` / `ContractCreate` / `Precompile` supported transactions.

Th mirrorNode parsing logic expects each transaction in the record stream to contain all the necessary persistence information in the combination of TransactionBody or the TransactionBody.
In come cases for security and performance some data is not repeated from the parent transaction to the child transaction.
This requires the MirrorNode to infer or set defaults.

To avoid this the MirrorNode should add logic to be able to refer to a previously parsed parent transaction for its details during the processing of a child transaction.

### Solution

From v0.26 of services the following assumptions can be made
- The parent and child transactions will be in the same recordFile
- The parent transaction will be sequentially followed by the child transactions

With that in mind the parser at most has to keep track of the parent transaction for n - 1 transactions.
Where n is the number of connected parent and child transactions.

To achieve this
- Add a `RecordItem` reference to the `RecordItem` class to contain the parentRecordItem. 
- Add logic to either the `RecordFileReader` or the `EntityRecordItemListener` to track the parentRecordItem. Add consideration to only update should the current object not be a child
- Update `ContractCreateTransactionHandler` with logic to check for child transaction scenario. Logic should manage both `EthereumTransaction` or `ContractCreate` as the parent transaction type, extracting the initCode/FileId for population in Contract

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Nana-EC**
Signed-off-by: Nana-EC <nana.essilfie-conduah@hedera.com>

**Description**:
Add logic to refer to a parent transaction in the parser
- Add RecordItem parent & previous links in `RecordItem`
- Update `FileReader` to populate previous and `RecordItem` to populate `parent` where the recordItem is a child.
- Update `ContractCreateTransactionHandler` to handle when initCode or FileID is in parent transaction, either `ContractCreate` or `EthereumTransactionHandler`
- Add `EthereumTransaction` containing rcd test files for coverage expansion

**Related issue(s)**:

Fixes #3670 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 1]
```diff
@@ -107,19 +108,39 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
-            items.add(recordStreamObject.getRecordItem());
+            var recordItem = recordStreamObject.getRecordItem();
+
+            // check if current item is a child
+            if (recordItem.isChild()) {
+                if (lastRecordItem != null &&
+                        DomainUtils.timestampInNanosMax(recordItem.getRecord().getParentConsensusTimestamp())
+                                .longValue() ==
+                                DomainUtils.timestampInNanosMax(lastRecordItem.getRecord().getConsensusTimestamp())
+                                        .longValue()) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:125]: Can be simplified and avoid the unnecessary timestamp conversion:

```suggestion
            if (recordItem.isChild() && parentRecordItem != null && recordItem.getParentConsensusTimestamp().equals(parentRecordItem.getConsensusTimestamp()) {
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 2]
```diff
@@ -107,19 +108,39 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
             }
 
             lastRecordStreamObject = recordStreamObject;
             count++;
+
+            if (!recordItem.isChild()) {
+                // update last recordItem reference for next item. Preserve parent until all children have been
+                // processed since they are assured to exist in sequential order of [Parent, Child1,...,ChildN]
+                lastRecordItem = recordItem;
+            }
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:143]: Move this to an else above so we don't check `isChild()` twice.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 3]
```diff
@@ -116,4 +124,51 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
...
         }
     }
+
+    public void updateChildFromParent(Contract contract, RecordItem recordItem) {
+        if (!entityProperties.getPersist().isContracts() || !recordItem.isChild()) {
+            return;
+        }
+
+        // parents may be ContractCreate or EthereumTransaction
+        var parentRecordItem = recordItem.getParentRecordItem();
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:134]: Check if parent is null and return. We can have pre-0.26 case where child and parent are in different record files.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 4]
```diff
@@ -116,4 +124,51 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
...
+        switch (TransactionType.of(parentRecordItem.getTransactionType())) {
+            case CONTRACTCREATEINSTANCE:
+                updateChildFromContractCreateParent(contract, parentRecordItem);
+                break;
+            case ETHEREUMTRANSACTION:
+                updateChildFromEthereumTransactionParent(contract, parentRecordItem);
+                break;
+            default:
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:142]: So if parent is contract call and child is contract create we'll never have the data?

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 5]
```diff
@@ -116,4 +124,51 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
...
+                // should we throw in this case?
+                break;
+        }
+    }
+
+    private void updateChildFromEthereumTransactionParent(Contract contract, RecordItem recordItem) {
+        var body = recordItem.getTransactionBody().getEthereumTransaction();
+        var ethereumDataBytes = DomainUtils.toBytes(body.getEthereumData());
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:165]: As an optimization, we should skip extracting bytes and parsing data if the `EthereumTransactionBody.callData` is present.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 6]
```diff
@@ -116,4 +124,51 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
...
+        var ethereumTransaction = ethereumTransactionParser.decode(ethereumDataBytes);
+
+        if (ethereumTransaction.getCallData() != null) {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:168]: It's not clear if call data will be null or an empty byte array if the FileID is not present. Regardless, we should definitely check FileID first then we can always set this as the fallback behavior and don't have to worry about null vs empty.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:134]: Should never happen as the timestamp check ensure they are both in the same file and therefore avoids pre v0.26
Can add it here for defensiveness though

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 4 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:142]: Yep.
That flow needs a migration to look up contract.file and set it.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:165]: Good point

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 3 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:134]: It can definitely happen. If parent is last item in RF1 and child is first item in RF2, then any timestamp check in record item reader won't occur and `recordItem.isChild()` will return true. NPE occurs

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:168]: Had actually locally changed this to set based on `contract` fields being null

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/transaction/RecordItem.java [hunk 7]
```diff
@@ -71,6 +72,9 @@ public class RecordItem implements StreamItem {
 
     private final Integer transactionIndex;
 
+    @NonFinal
+    private RecordItem parentRecordItem = null;
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/transaction/RecordItem.java]: Changed

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 8]
```diff
@@ -107,19 +108,39 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
         long consensusStart = 0;
         RecordStreamObject lastRecordStreamObject = null;
         List<RecordItem> items = new ArrayList<>();
+        RecordItem lastRecordItem = null;
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 9]
```diff
@@ -107,19 +108,39 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
+            var recordItem = recordStreamObject.getRecordItem();
+
+            // check if current item is a child
+            if (recordItem.isChild()) {
+                if (lastRecordItem != null &&
+                        DomainUtils.timestampInNanosMax(recordItem.getRecord().getParentConsensusTimestamp())
+                                .longValue() ==
+                                DomainUtils.timestampInNanosMax(lastRecordItem.getRecord().getConsensusTimestamp())
+                                        .longValue()) {
+                    recordItem.setParentRecordItem(lastRecordItem);
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Problem is we need to parse the recordBytes first and create the child recordItem.
Then using that determine if setting a parentRecordItem is applicable.
Thats why I added the setter

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 1 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 2 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Moved

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 5 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:176]: Added, also added a test to verify

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 6 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 3 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Still don't see this check. Not sure if you've pushed the change.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 10]
```diff
@@ -107,12 +107,26 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
+
+            // check if current item is a child
+            if (recordItem.isChild() && parentRecordItem != null &&
+                    recordItem.getRecord().getParentConsensusTimestamp()
+                            .equals(parentRecordItem.getRecord().getConsensusTimestamp())) {
+                recordItem.setParent(parentRecordItem);
+            } else {
+                // update last recordItem reference for next item. Preserve parent until all children have been
+                // processed since they are assured to exist in sequential order of [Parent, Child1,...,ChildN]
+                parentRecordItem = recordItem;
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:126]: This would mark a child that's the first entry in a record file as being a parent.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Provided an alternative algorithm via slack

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 10 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:126]: Meant to nest iniital if.
Should be
```java
            if (recordItem.isChild()) {
                if (parentRecordItem != null &&
                        recordItem.getRecord().getParentConsensusTimestamp()
                                .equals(parentRecordItem.getRecord().getConsensusTimestamp())) {
                    recordItem.setParent(parentRecordItem);
                }
            } else {
                // update last recordItem reference for next item. Preserve parent until all children have been
                // processed since they are assured to exist in sequential order of [Parent, Child1,...,ChildN]
                parentRecordItem = recordItem;
            }
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Implemented, with tests

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 3 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Missed it, added now

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 10 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 11]
```diff
@@ -105,27 +105,35 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
+        RecordItem lastRecordItem = null;
 
         // read record stream objects
         while (!isHashObject(vdis, hashObjectClassId)) {
             RecordStreamObject recordStreamObject = new RecordStreamObject(vdis, recordFile.getHapiVersion(),
                     count);
-            items.add(recordStreamObject.getRecordItem());
+            var recordItem = RecordItem.builder()
+                    .hapiVersion(recordFile.getHapiVersion())
+                    .previous(lastRecordItem)
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:117]: Since now we're setting previous in addition to parent, we could set it on v1 and v2 readers since otherwise our future selves might try to use `RecordItem.getPrevious()` and get a NPE on historical data.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java [hunk 12]
```diff
@@ -105,27 +105,35 @@ private void readBody(ValidatedDataInputStream vdis, DigestInputStream metadataD
...
-        RecordStreamObject lastRecordStreamObject = null;
         List<RecordItem> items = new ArrayList<>();
+        RecordItem lastRecordItem = null;
 
         // read record stream objects
         while (!isHashObject(vdis, hashObjectClassId)) {
             RecordStreamObject recordStreamObject = new RecordStreamObject(vdis, recordFile.getHapiVersion(),
                     count);
-            items.add(recordStreamObject.getRecordItem());
+            var recordItem = RecordItem.builder()
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: We should add a test in `RecordFileReaderTest`.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: I looked into that and it's quite heavy since this method and others aren't easily exposed. Seems like there would be multiple levels of mocking to get to this depth.
Also with the new logic there's not much change to this file in terms of logic so the existing tests should suffice.
We now just create and cache the RecordItem that was done in the streamObject directly instead of calling StreatObject.getRecordItem multiple times.

Actually i should remove `getRecordItem()` since that's dead code now

Let me know if there's a flow in here that you think was changed that is of concern

Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: I'm not sure we're talking about the same thing. I meant add a test to `RecordFileReaderImplV5Test` that verifies that previous is not null and is the expected item and that parent is not null. If you add it to V1 and V2 then you can pull up to `RecordFileReaderTest`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 13]
```diff
@@ -116,4 +124,64 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
             contractResult.setGasLimit(transactionBody.getGas());
         }
     }
+
+    public void updateChildFromParent(Contract contract, RecordItem recordItem) {
+        if (!entityProperties.getPersist().isContracts() || !recordItem.isChild() || recordItem.getParent() == null) {
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:129]: can be simplified to 
```suggestion
        if (recordItem.getParent() == null) {
```

`isContracts()` is checked in the caller

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 14]
```diff
@@ -116,4 +124,64 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
...
+            contract.setFileId(EntityId.of(body.getCallData()));
+            return;
+        }
+
+        var ethereumDataBytes = DomainUtils.toBytes(body.getEthereumData());
+        var ethereumTransaction = ethereumTransactionParser.decode(ethereumDataBytes);
+
+        if (contract.getFileId() == null) {
+            contract.setFileId(ethereumTransaction.getCallDataId());
+        }
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:181]: can remove, this is dead code, `etheremTransaction.callDataId` is always `null`

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java [hunk 15]
```diff
@@ -195,6 +205,187 @@ void updateTransactionSuccessfulWithEvmAddressAndInitcode() {
...
+                .build();
+        var contractId = EntityId.of(recordItem.getRecord().getReceipt().getContractID());
+        var timestamp = recordItem.getConsensusTimestamp();
+        var transaction = domainBuilder.transaction()
+                .customize(t -> t.consensusTimestamp(timestamp).entityId(contractId))
+                .get();
+        transactionHandler.updateTransaction(transaction, recordItem);
+        assertContract(contractId, timestamp, t -> assertThat(t)
+                .returns(null, Contract::getFileId)
+                .satisfies(c -> assertThat(c.getInitcode()).isNotNull())
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:280]:
```suggestion
                .satisfies(c -> assertThat(c.getInitcode()).isNotEmpty())
```

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java [hunk 16]
```diff
@@ -116,4 +124,64 @@ public void updateContractResult(ContractResult contractResult, RecordItem recor
             contractResult.setGasLimit(transactionBody.getGas());
         }
     }
+
+    public void updateChildFromParent(Contract contract, RecordItem recordItem) {
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:128]:
```suggestion
    private void updateChildFromParent(Contract contract, RecordItem recordItem) {
```

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java [hunk 17]
```diff
@@ -195,6 +205,187 @@ void updateTransactionSuccessfulWithEvmAddressAndInitcode() {
...
+        var parentRecordItem = recordItemBuilder.contractCreate()
+                .transactionBody(x -> x.clearFileID()
+                        .setInitcode(ByteString.copyFrom("init code", StandardCharsets.UTF_8)))
+                .build();
+
+        // child item
+        var recordItem = recordItemBuilder.contractCreate()
+                .transactionBody(b -> b.clearFileID().clearInitcode())
+                .parent(parentRecordItem)
+                .build();
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:220]: why would a recordItem with a transaction record (parentConsensusTimestamp = null) have a parentRecordItem?

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java [hunk 18]
```diff
@@ -195,6 +205,187 @@ void updateTransactionSuccessfulWithEvmAddressAndInitcode() {
...
+        );
+    }
+
+    @Test
+    void updateContractFromEthereumTransactionWCallDataFileParent() {
+        // parent item
+        var parentRecordItem = recordItemBuilder.ethereumTransaction(true)
+                .build();
+
+        doReturn(domainBuilder.ethereumTransaction(true).get()).when(ethereumTransactionParser).decode(any());
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:369]:
```suggestion
```

`decode()` isn't called

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java [hunk 19]
```diff
@@ -195,6 +205,187 @@ void updateTransactionSuccessfulWithEvmAddressAndInitcode() {
...
+        );
+    }
+
+    @Test
+    void updateContractFromEthereumTransactionWFileIDParent() {
+        // parent item
+        var parentRecordItem = recordItemBuilder.ethereumTransaction(true)
+                .build();
+
+        doReturn(domainBuilder.ethereumTransaction(false).get()).when(ethereumTransactionParser).decode(any());
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:343]: the stubbed return doesn't match the implementation since `EtherumTransactionParser` never sets `callDataId`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 12 above)
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: remove `transactionBlockIndex` from `RecordStreamObject`, remove `RecordStreamObject.getRecordItem`

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: Yep, removed locally, coming in next commit

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 13 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java:129]: Yeah, I dropped one of the tests after realizing this.
Conceptually I see why it can be reduced further, however `recordItem.isChild()` is set as a result of transactionRecordBytes parsing, whereas `recordItem.getParent()` is set as a result of comparison logic between current and previous.
Because the logic exists in different flows I think it's fair to have the explicit checks.
Let me know if you disagree

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java (hunk 17 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:220]: Similar comment to other. Just exhaustive testing given the unique nature contributing to each items set.
In practice it shouldn't happen but if an inconsistency did occur we have coverage for it's behaviour

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java (hunk 18 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java:369]: `decode` is always called if of `EthereumTransaction`.
Let me know if i'm missing something

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/transaction/RecordItem.java [hunk 20]
```diff
@@ -71,6 +71,10 @@ public class RecordItem implements StreamItem {
 
     private final Integer transactionIndex;
 
+    private final RecordItem parent;
+
+    private RecordItem previous;
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/transaction/RecordItem.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 11 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:117]: Added

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: Added tests.
I don't believe there was a parent-child relationship preV5 so the tests for parent had to be in v5 but the tests for previous is shared

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 14 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Removed

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java (hunk 15 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java]: Updated

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 16 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: Updated

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java (hunk 19 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java]: That's correct. It's set in the `upTransaction()` formerly `getEntity`.
The stub is to ensure the parser call result is successful.
I can explicitly set the `callDataId` to null although as you said it's not picked up there.
Rather `callDataId` is always pulled from the `transactionBody`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java (hunk 12 above)
Comment (Nana-EC) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/reader/record/RecordFileReaderImplV5.java:115]: Removed

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java (hunk 13 above)
Comment (xin-hedera) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandler.java]: I'd say checking parent set or not is enough, since if it's not a child there definitely shouldn't be parent set. if in any chance the parent is set, that's a bug in `RecordItem`, not here.

File: hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java (hunk 18 above)
Comment (xin-hedera) [reply hedera-mirror-importer/src/test/java/com/hedera/mirror/importer/parser/record/transactionhandler/ContractCreateTransactionHandlerTest.java]: the problem is `decode` isn't always called. the fast path in the implementation is if the call data is stored in a file, the transaction handler just gets the file id from the tx body, so ethereum tx parsing is skipped.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by xin-hedera — APPROVED** [review]
LGTM
